Expand Boolean API convention example (#1224)
openshift/enhancements: dev-guide/api-conventions.md

While the upstream Kubernetes conventions recommend thinking twice about using Booleans, they are explicitly forbidden
within OpenShift APIs.

Many ideas start as a Boolean value, e.g. `FooEnabled: true|false`, but often evolve into needing 3, 4, or even more
states at some point during the API's lifetime.
As a Boolean value can only ever have 2 or in some cases 3 values (`true`, `false`, `omitted` when a pointer), we have

> seen examples in which API authors have later added additional fields, paired with a Boolean field that are only

Reviewer: ```suggestion
seen examples in which API authors have later added additional fields, paired with a Boolean field, that are only
```

meaningful when the original field has a certain state. This makes it confusing for an end user as they have to be

> aware that the field they are trying to use, only has an effect in certain circumstances.

Reviewer: ```suggestion
aware that the field they are trying to use only has an effect in certain circumstances.
```

Use an enumeration of values that describe the action instead:
```go
// authentication determines the requirements for authentication within the cluster. Allowed values are "Optional",
// "Required", "Disabled" and omitted.

> // Optional authentication allows users to optionally authenticate but will not reject an unathenticated request.
> // Required authentication requires all requests to be authenticated.
> // Disabled authentication ignores any attempt to authenticate and proccesses all requests as unathenticated.

Reviewer: ```suggestion
// Optional authentication allows users to optionally authenticate but will not reject an unauthenticated request.
// Required authentication requires all requests to be authenticated.
// Disabled authentication ignores any attempt to authenticate and processes all requests as unauthenticated.
```
